File: data/EXP01/python/2556/2556.output_3.py
# @lc app=leetcode id=2556 lang=python3
from typing import List

class Solution:
    def dfs(self, grid: List[List[int]], visited: List[List[bool]], row: int, col: int) -> None:
        m, n = len(grid), len(grid[0])
        
        if row < 0 or row >= m or col < 0 or col >= n or visited[row][col] or grid[row][col] == 0:
            return
        
        visited[row][col] = True
        
        self.dfs(grid, visited, row + 1, col)
        self.dfs(grid, visited, row, col + 1)
        
        
    def isPossibleToCutPath(self, grid: List[List[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        
        visited = [[False] * n for _ in range(m)]
        
        self.dfs(grid, visited, 0, 0)
        
        if visited[m - 1][n - 1]:
            return False
        
        for row in range(m):
            for col in range(n):
                grid[row][col] = 1 - grid[row][col]
                
                visited = [[False] * n for _ in range(m)]
                
                self.dfs(grid, visited, 0, 0)
                
                if not visited[m - 1][n - 1]:
                    return True
                
                grid[row][col] = 1 - grid[row][col]
                
        return False

File: data/EXP01/go/2556/2556.output_3.go
func isPossibleToCutPath(grid [][]int) bool {
    m := len(grid)
    n := len(grid[0])
  
    if !hasPath(grid, 0, 0, m-1, n-1) {
        return true
    }
  
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            temp := grid[i][j]
            grid[i][j] ^= 1
  
            if !hasPath(grid, 0, 0, m-1, n-1) {
                return true
            }
  
            grid[i][j] = temp
        }
    }
  
    return false
}

func hasPath(grid [][]int, r1, c1, r2, c2 int) bool {
    if grid[r1][c1] == 0 || grid[r2][c2] == 0 {
        return false
    }
  
    m := len(grid)
    n := len(grid[0])
  
    visited := make([]bool, m*n)
    queue := []int{r1*n + c1}
    visited[r1*n+c1] = true
  
    for len(queue) > 0 {
        size := len(queue)
  
        for i := 0; i < size; i++ {
            curr := queue[i]
            r := curr / n
            c := curr % n
  
            if r == r2 && c == c2 {
                return true
            }
  
            if r < m-1 && grid[r+1][c] == 1 && !visited[(r+1)*n+c] {
                queue = append(queue, (r+1)*n+c)
                visited[(r+1)*n+c] = true
            }
  
            if c < n-1 && grid[r][c+1] == 1 && !visited[r*n+c+1] {
                queue = append(queue, r*n+c+1)
                visited[r*n+c+1] = true
            }
        }
  
        queue = queue[size:]
    }
  
    return false
}
